libr/asm/p/asm_mips_gnu.c

```diff
@@ -43,8 +45,37 @@ static int disassemble(struct r_asm_t *a, struct r_asm_op_t *op, const ut8 *buf,
 	Offset = a->pc;
 	memcpy (bytes, buf, 4); // TODO handle thumb
 
+	if ((a->cpu != pre_cpu) && (a->features != pre_features)) {
+		free (disasm_obj.disassembler_options);
+		memset (&disasm_obj, '\0', sizeof (struct disassemble_info));
+	}	
+
 	/* prepare disassembler */
-	memset (&disasm_obj,'\0', sizeof (struct disassemble_info));
+	if (a->cpu && a->cpu != pre_cpu) {
+		if (!r_str_casecmp (a->cpu, "mips64r2")) {
+			disasm_obj.mach = bfd_mach_mipsisa64r2;
+		} else if (!r_str_casecmp (a->cpu, "mips32r2")) {
+			disasm_obj.mach = bfd_mach_mipsisa32r2;
+		} else if (!r_str_casecmp (a->cpu, "mips64")) {
+			disasm_obj.mach = bfd_mach_mipsisa64;
+		} else if (!r_str_casecmp (a->cpu, "mips32")) {
+			disasm_obj.mach = bfd_mach_mipsisa32;
+		}
+		pre_cpu = a->cpu;
+	}
+
+	if (a->features && a->features != pre_features) {
+		disasm_obj.disassembler_options = r_str_new("");
```

Comment:
missing space before (, and as i said, this can be = NULL; the code below should work without any change

Reply:
I tried but got some error. 

 882 R_API char *r_str_append(char *ptr, const char *string) {
 883     if (string && !ptr) {
 884         **return strdup (string);**
 885     }
 886     if (!string) {
 887         return ptr;
 888     }

Reply:
which error did you get? if ptr is null it will return strdup(string), looks correct to me

Reply:
ah well. did you set the pointer returned? like 
```
char *do = &disasm_obj.disassembler_options;
*do = r_str_append (*do, "abi=n64");
```

Reply:
This is probably the reason why its segfaulting on bsd and windows